tasks/linux.sh

```diff
@@ -1,61 +1,162 @@
+
 action="$PT_action"
 name="$PT_name"
 version="$PT_version"
+package_managers=("apt-get" "yum")
+options=()
 
-package_managers[0]="yum"
-package_managers[1]="apt-get"
-
-# example cli /opt/puppetlabs/puppet/bin/bolt task run package::linux action=install name=vim --nodes localhost --modulepath /etc/puppetlabs/code/modules --password puppet --user root
+for p in "${package_managers[@]}"; do
+  if type "$p" &>/dev/null; then
+    available_manager="$p"
+    break
+  fi
+done
 
-check_command_exists() {
-  (which "$1") > /dev/null 2>&1
-  command_exists=$?
-  return ${command_exists}
+[[ $available_manager ]] || {
+  validation_error "No package managers found: Must be one of: [apt, yum]"
 }
 
-if [ "${action}" = "uninstall" ]; then
-  action="remove"
-fi
-
-for package_manager in "${package_managers[@]}"
-do
-  check_command_exists "$package_manager"
-  command_exists=$?
-  if [ "${command_exists}" -eq 0 ]; then
-    if [ "${package_manager}" = "yum" ]; then
-      if [ "${version}" != "" ]; then
-        name="$name-$version"
-      fi
-      # upgrading to a specific version needs to use the install action
-      if [ "${version}" != "" ] && [ "${action}" = "upgrade" ]; then
+# For any package manager, check if the action is "status". If so, only run a status command
+# Otherwise, run the requested action and follow up with a "status" command
+case "$available_manager" in
+  "apt-get")
+    # quiet and assume yes
+    options+=("-yq")
+
+    # <package>=<version> is the syntax for installing a specific version in apt
+    [[ $version ]] && name="${name}=${version}"
+
+    # Use --showformat to make a json object with status and version
+    # ${name%%=*} removes the version in case we installed a specific one
+    # This will print nothing to stdout if the package is not installed
+    case "$action" in
+      "uninstall")
+        action="remove"
+        ;;
+      "upgrade")
         action="install"
-      fi
+        options+=("--only-upgrade")
+        old_version="$(dpkg-query --show --showformat='"${Version}"' ${name%%=*})"
+    esac
+
+    [[ $action == "status" ]] || DEBIAN_FRONTEND=noninteractive "apt-get" "$action" "$name" "${options[@]}" >/dev/null || fail
+
+    # uninstall: do not print any version information
+    if [ "$PT_action" == 'uninstall' ]; then
+      cmd_status="$(dpkg-query --show --showformat='{ "status":"${Status}" }' ${name%%=*})"
+      [[ $cmd_status ]] || cmd_status='{ "status": "uninstalled"}'
+      success "$cmd_status"
+    fi
+
+    # upgrade: only show version information
+    if [ "$PT_action" == 'upgrade' ]; then
+      new_version="$(dpkg-query --show --showformat='"${Version}"' ${name%%=*})"
+      cmd_status="{ \"old_version\":"${old_version}", \"version\":"${new_version}" }"
+      success "$cmd_status"
+    fi
+
+    # Determine if newer package is available to mirror the ruby/puppet implementation
+    upgradable_list=($(sudo apt-get -u -V --assume-no upgrade | grep "\\${PT_name}\\b"))
```

Comment:
This could probably be simplified to just check the output of `apt-get` with `grep` and use the return code.  Also, a specific package can be given to `apt-get`.  For example, with an up-to-date `puppet-agent`

```
$cat foo.sh
#!/bin/bash

if sudo apt-get --assume-no upgrade puppet-agent | grep -q 'already the newest version'; then
  echo "latest"
else
  echo "upgradable"
fi
```
```
./foo.sh
latest
```